R/develop-request.R

```diff
@@ -0,0 +1,218 @@
+#' \dontrun{
+#' ## Example with a prepared endpoint
+#' ept <- googledrive::drive_endpoints("drive.files.update")[[1]]
+#' req <- develop_request(
+#'   ept,
+#'   params = list(
+#'     fileId = "abc",
+#'     addParents = "123",
+#'     description = "Exciting File"
+#'   )
+#' )
+#' req
+#'
+#' req <- build_request(
+#'   method = req$method,
+#'   path = req$path,
+#'   params = req$params,
+#'   body = req$body,
+#'   token = "PRETEND_I_AM_A_TOKEN"
+#' )
+#' req
+#'
+#' ## Example with no previous knowledge of the endpoint
+#' ## List a file's comments
+#' ## https://developers.google.com/drive/v3/reference/comments/list
+#' req <- build_request(
+#'   method = "GET",
+#'   path = "drive/v3/files/{fileId}/comments",
+#'   params = list(
+#'     fileId = "your-file-id-goes-here",
+#'     fields = "*"
+#'   ),
+#'   token = "PRETEND_I_AM_A_TOKEN"
+#' )
+#' req
+#' }
+develop_request <- function(endpoint,
+                            params = list(),
+                            base_url = "https://www.googleapis.com") {
+  check_params(params, endpoint$parameters)
+  body_params <- Filter(function(x) x$location == "body", endpoint$parameters)
+  params <- partition_params(params, names(body_params))
+  list(
+    method = endpoint$method,
+    path = endpoint$path,
+    params = params$unmatched,
+    body = params$matched,
+    base_url = base_url
+  )
+}
+
+#' @rdname develop_request
+#' @export
+build_request <- function(method = "GET",
+                          path = "",
+                          params = list(),
+                          body = list(),
+                          token = NULL,
+                          key = NULL,
+                          base_url = "https://www.googleapis.com") {
+
+  params <- partition_params(params, extract_path_names(path))
+
+  ## send a token or a key, but never both
+  params$unmatched$key <- if (is.null(token)) {
+    key %||% params$unmatched$key
+  } else {
+    NULL
+  }
+
+  out <- list(
+    method = method,
+    path = glue::glue_data(params$matched, path),
+    query = params$unmatched,
+    body = body,
+    token = token
+  )
+
+  out$url <- httr::modify_url(
+    url = base_url,
+    path = out$path,
+    ## prevent a trailing `?` or `?=` when the query is trivial, e.g. list() or
+    ## contains a single element which is NULL
+    ## https://github.com/r-lib/httr/issues/451
+    ## could simply use out$query this if required httr >= 1.3.1
+    query = if (length(unlist(out$query)) == 0) NULL else out$query
+  )
+  out
+}
+
+## check params provided by user against spec
+##   * error if required params are missing
+##   * error for unknown params
+check_params <- function(provided, spec) {
+  required <- Filter(function(x) isTRUE(x$required), spec)
+  missing <- setdiff(names(required), names(provided))
+  if (length(missing)) {
+    stop_collapse(
+      c("Required parameter(s) are missing:", missing)
+    )
+  }
+
+  unknown <- setdiff(names(provided), names(spec))
```

Comment:
Do you need to only do this if `additionalProperties: true` isn't in the disco doc?

Reply:
This is my first contact with `"additionalProperties"`, so I will educate myself about that.

Reply:
http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.5.6

Reply:
Relevant if we went this far (which I propose we do not, at this time):

https://cran.r-project.org/web/packages/jsonvalidate/vignettes/jsonvalidate.html